engine/inc/com/centreon/engine/configuration/applier/pb_difference.hh

```diff
@@ -0,0 +1,147 @@
+/**
+ * Copyright 2023 Centreon
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For more information : contact@centreon.com
+ */
+
+#ifndef CCE_CONFIGURATION_APPLIER_PB_DIFFERENCE_HH
+#define CCE_CONFIGURATION_APPLIER_PB_DIFFERENCE_HH
+
+#include <absl/container/flat_hash_map.h>
+#include <absl/container/flat_hash_set.h>
+#include <google/protobuf/util/message_differencer.h>
+
+#include <iterator>
+
+namespace com::centreon::engine {
+
+using MessageDifferencer = ::google::protobuf::util::MessageDifferencer;
+
+namespace configuration {
+namespace applier {
+template <typename T,
+          typename Key,
+          typename Container = ::google::protobuf::RepeatedPtrField<const T>>
+class pb_difference {
+  // What are the new objects
+  std::vector<T> _added;
+  // What index to delete
+  std::vector<std::pair<ssize_t, Key>> _deleted;
+  // A vector of pairs, the pointer to the old one and the new one.
+  std::vector<std::pair<T*, T>> _modified;
+
+ public:
+  /**
+   * @brief Default constructor.
+   */
+  pb_difference() = default;
+
+  /**
+   * @brief Destructor.
+   */
+  ~pb_difference() noexcept = default;
+  pb_difference(const pb_difference&) = delete;
+  pb_difference& operator=(const pb_difference&) = delete;
+  const std::vector<T>& added() const noexcept { return _added; }
+  const std::vector<std::pair<ssize_t, Key>>& deleted() const noexcept {
+    return _deleted;
+  }
+  const std::vector<std::pair<T*, T>>& modified() const noexcept {
+    return _modified;
+  }
+
+  template <typename Function>
+  void parse(typename Container::iterator old_first,
+             typename Container::iterator old_last,
+             typename Container::iterator new_first,
+             typename Container::iterator new_last,
+             Function f) {
+    absl::flat_hash_map<Key, T*> keys_values;
```

Comment:
it is possible to use absl::flat_hash_map<Key, const T*> instead in order to avoid const_cast?